compiler/src/dmd/location.d

```diff
@@ -46,8 +69,8 @@ nothrow:
     }
 
     extern (C++) const(char)* toChars(
-        bool showColumns = global.params.showColumns,
-        ubyte messageStyle = global.params.messageStyle) const pure nothrow
+        bool showColumns = Loc.showColumns,
+        ubyte messageStyle = Loc.messageStyle) const pure nothrow
```

Comment:
I expect the testsuite to fail.

Reply:
> I expect the testsuite to fail.

If it is doesn't, then you've successfully circumvented the C++ interface tests.

Reply:
I don't see why.

Reply:
> > I expect the testsuite to fail.
> 
> If it is doesn't, then you've successfully circumvented the C++ interface tests.

Because gdc and ldc _will_ break as a result of this change until you have updated the headers too.

Reply:
The first thing I spot is that these fields don't appear to be removed from the params struct.

Reply:
@ibuclaw ping

Reply:
Oh I see... C++ doesn't encode the default initializer into the signature. So there would be a split brain between dmd and gdc/ldc (the c++ interface), where the later will still call this passing global.params...

That this will break downstream still stands though.  As setting the message style will be ignored by the D code because it checks a different global.

Reply:
> these fields don't appear to be removed from the params struct.

They stay in the params struct because the purpose of the params struct is to collect all the user-supplied configuration.

> there would be a split brain between dmd and gdc/ldc (the c++ interface), where the latter will still call this passing global.params...

I don't understand. The only references to global.params.showColumns/messageStyle are in mars.d where they are set by the user interface and then passed to `Loc.set()`.

> That this will break downstream still stands though. As setting the message style will be ignored by the D code because it checks a different global, and I don't see the setter exposed anywhere.

`Loc.set()` is a public function exposed by `dmd.location`. I'm not sure exactly what you want. Please be more specific.

Reply:
@ibuclaw ping

Reply:
> > there would be a split brain between dmd and gdc/ldc (the c++ interface), where the latter will still call this passing global.params...
> 
> I don't understand. The only references to global.params.showColumns/messageStyle are in mars.d where they are set by the user interface and then passed to `Loc.set()`.
> 

They should be set by the c++ interface tests too, and if not, then that is an oversight, as that should be representative of what gdc/ldc do.

Mars.d is not used by gdc and ldc, `Loc::set()` can't be called because it's not in the c++ interface headers.

> > That this will break downstream still stands though. As setting the message style will be ignored by the D code because it checks a different global, and I don't see the setter exposed anywhere.
> 
> `Loc.set()` is a public function exposed by `dmd.location`. I'm not sure exactly what you want. Please be more specific.

1. It's not made public to c++ (ie: calling `Loc::set()` from the c++ interface tests will result in an undeclared function compiler error.

2. The D definition of `Loc.toChars` no longer matches the [C++ declaration](https://github.com/dlang/dmd/blob/4b8496192b647754a85c84f02e46f627feb7047f/compiler/src/dmd/globals.h#L359).

Reply:
what file do you want the test in?

Reply:
@ibuclaw ping

Reply:
@ibuclaw ping

Reply:
This is the function that tests the Loc interface from C++.
https://github.com/dlang/dmd/blob/c5d192dfa3bab798f5adbdb8b87d83906c6e99b1/compiler/src/tests/cxxfrontend.cc#L412-L418

This ought to cover it - looks like this would work now.
```
     assert(strcmp(loc1.toChars(), "test.d(24)") == 0);
     Loc::set(true, MESSAGESTYLEdigitalmars);
     assert(strcmp(loc1.toChars(), "test.d(24,42)") == 0);
     Loc::set(false, MESSAGESTYLEgnu);
     assert(strcmp(loc1.toChars(), "test.d:24") == 0);
```